coap/adapter.go

```diff
@@ -1,61 +1,257 @@
-package adapter
+package coap
 
 import (
 	"encoding/json"
-	"log"
+	"errors"
+	"fmt"
 	"net"
+	"strings"
 
-	"github.com/dustin/go-coap"
+	"encoding/base64"
+
+	mux "github.com/dereulenspiegel/coap-mux"
+	coap "github.com/dustin/go-coap"
+	"github.com/go-kit/kit/log"
 	"github.com/mainflux/mainflux"
+	manager "github.com/mainflux/mainflux/manager/client"
 	broker "github.com/nats-io/go-nats"
-	"go.uber.org/zap"
 )
 
-const protocol string = "coap"
+const (
+	token    string = "token"
+	prefix   string = "src.coap."
+	channel  string = "channel_id"
+	protocol string = "coap"
+)
+
+var (
+	errUnauthorizedAccess error = errors.New("missing or invalid credentials provided")
+)
 
+// Observer struct provides support to observe message types.
 type Observer struct {
 	conn    *net.UDPConn
 	addr    *net.UDPAddr
 	message *coap.Message
+	sub     *broker.Subscription
 }
 
-type CoAPAdapter struct {
+// Subscriber represents token-NATS subscription map. Token is represented as string in
+// order to simplify access to subscription as well as to simplify unsubscribe feature.
+type Subscriber map[string]*Observer
+
+// AdapterService struct represents CoAP adapter service.
+type AdapterService struct {
 	obsMap map[string][]Observer
-	logger *zap.Logger
+	logger log.Logger
 	pub    mainflux.MessagePublisher
+	subs   map[string]Subscriber
+	nc     *broker.Conn
+	mc     *manager.ManagerClient
 }
 
-// NewCoAPAdapter creates new CoAP adapter struct
-func NewCoAPAdapter(logger *zap.Logger, pub mainflux.MessagePublisher) *CoAPAdapter {
-	ca := &CoAPAdapter{
+// New creates new CoAP adapter service struct.
+func New(logger log.Logger, pub mainflux.MessagePublisher, nc *broker.Conn, mc *manager.ManagerClient) *AdapterService {
+	ca := &AdapterService{
 		logger: logger,
 		pub:    pub,
 		obsMap: make(map[string][]Observer),
+		subs:   make(map[string]Subscriber),
+		nc:     nc,
+		mc:     mc,
 	}
-
 	return ca
 }
 
-// Serve function starts CoAP server
-func (ca *CoAPAdapter) Serve(addr string) error {
-	ca.logger.Info("Starting CoAP server", zap.String("address", addr))
-	return coap.ListenAndServe("udp", addr, ca.COAPServer())
+// SendMessage method sends processes message and pushes it to NATS.
+func (ca *AdapterService) SendMessage(l *net.UDPConn, a *net.UDPAddr, m *coap.Message) *coap.Message {
+	ca.logger.Log("info", fmt.Sprintf("Got message in sendMessage: path=%q: %#v from %v", m.Path(), m, a))
+	var res *coap.Message
+	if len(m.Payload) == 0 && m.IsConfirmable() {
+		res.Code = coap.BadRequest
+		return res
+	}
+
+	// Uri-Query option ID is 15 (0xf).
+	token, err := ca.token(m.Option(0xf))
```

Comment:
What's this token here? Is it CoAP token or some auth bearer token?

Because CoAP `token` is just a request ID: https://tools.ietf.org/html/rfc7252#section-5.3.1

Reply:
Its JWT token. JWT token is sent as url-query param, which is embeded in CoAP message options, with reserved code 15. Sorry for messy code, it's WIP version.